File: patch-tool/patch.js
#!/usr/bin/env node

const crypto = require('crypto')
const fs = require('fs')
const path = require('path')
const os = require('os')

const DEFAULT_API_URL = 'https://server.codeium.com'
const CONFIG_KEY = 'codeium.apiServerUrl'
const REGISTER_CONFIG_KEY = 'codeium.registerApiServerUrl'
const DEFAULT_REGISTER_URL = 'https://register.windsurf.com'
const LEGACY_GATEWAY_PLACEHOLDER_API_KEY = 'sk-ws-01-gateway-placeholder'
const GATEWAY_PLACEHOLDER_PREFIX = 'sk-ws-01-client-'
const ACP_CONFIG_KEY = 'windsurf.acp.enabledAgents'
const PATCH_STATE_PATH = path.join('User', 'globalStorage', 'windsurf-gateway-patch.json')
const AUTH_SESSION_FALLBACK_MARKER = 'account:{label:"Gateway",id:"windsurf-gateway"},scopes:[]}'
const AUTH_SESSION_FALLBACK_BLOCK_REGEX = /\?\?\{id:"windsurf-gateway",accessToken:"[^"]+",account:\{label:"Gateway",id:"windsurf-gateway"\},scopes:\[\]\}/
const AUTH_SESSION_FALLBACK_REGEX = /await i\.authentication\.getSession\(n\.WindsurfExtensionMetadata\.getInstance\(\)\.authProviderId,\[[\s\S]*?\],e\)/
const USER_STATUS_FALLBACK_SENTINEL = `allowedCommandModelConfigsProtoBinaryBase64:[],userStatusProtoBinaryBase64:""}`
const USER_STATUS_FALLBACK_REGEX = /([A-Za-z_$][\w$]*)\.StatusBar\.getInstance\(\)\.setAuthStatus\(!1\),([A-Za-z_$][\w$]*)\.windsurfAuth\.setAuthStatus\(null\),\(await\(0,([A-Za-z_$][\w$]*)\.getAuthSession\)\(\)\)\?\.accessToken===([A-Za-z_$][\w$]*)\|\|([A-Za-z_$][\w$]*)\.clearAuthentication\(\),!1/

function argValue(name) {
  const args = process.argv.slice(2)
  const eq = args.find(a => a.startsWith(`${name}=`))
  if (eq) return eq.slice(name.length + 1)
  const idx = args.indexOf(name)
  return idx >= 0 ? args[idx + 1] : null
}

function hasArg(name) {
  return process.argv.includes(name)
}

function homeConfigDir() {
  if (process.env.WINDSURF_CONFIG_DIR) return process.env.WINDSURF_CONFIG_DIR
  if (process.platform === 'darwin') return path.join(os.homedir(), 'Library', 'Application Support', 'Windsurf')
  if (process.platform === 'win32') return path.join(process.env.APPDATA || path.join(os.homedir(), 'AppData', 'Roaming'), 'Windsurf')
  return path.join(os.homedir(), '.config', 'Windsurf')
}

function defaultInstallDir() {
  if (process.env.WINDSURF_INSTALL_DIR) return process.env.WINDSURF_INSTALL_DIR
  if (process.platform === 'darwin') return '/Applications/Windsurf.app/Contents/Resources/app'
  if (process.platform === 'win32') return path.join(process.env.LOCALAPPDATA || '', 'Programs', 'Windsurf', 'resources', 'app')
  return '/opt/windsurf/resources/app'
}

const configDir = homeConfigDir()
const installDir = defaultInstallDir()
const settingsPath = path.join(configDir, 'User', 'settings.json')
const globalStatePath = path.join(configDir, 'User', 'globalStorage', 'state.vscdb')
const extensionPath = path.join(installDir, 'extensions', 'windsurf', 'dist', 'extension.js')
const backupRoot = path.join(configDir, 'windsurf-gateway-backups')
const patchStateFile = path.join(configDir, PATCH_STATE_PATH)

function banner() {
  console.log('\nWindsurf Gateway Patch Tool\n')
}

function backup(filePath) {
  if (!fs.existsSync(filePath)) return null
  const stamp = new Date().toISOString().replace(/[:.]/g, '-')
  const dir = path.join(backupRoot, stamp)
  fs.mkdirSync(dir, { recursive: true })
  const out = path.join(dir, path.basename(filePath))
  fs.copyFileSync(filePath, out)
  console.log(`backup ${filePath} -> ${out}`)
  return out
}

function readJson(filePath) {
  if (!fs.existsSync(filePath)) return {}
  return JSON.parse(fs.readFileSync(filePath, 'utf8') || '{}')
}

function writeJson(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true })
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + '\n')
}

function loadPatchState() {
  if (!fs.existsSync(patchStateFile)) return {}
  try {
    return JSON.parse(fs.readFileSync(patchStateFile, 'utf8') || '{}')
  } catch (e) {
    console.log(`patch state unreadable, regenerating: ${patchStateFile}`)
    return {}
  }
}

function savePatchState(state) {
  writeJson(patchStateFile, state)
}

function generatePlaceholderApiKey() {
  return GATEWAY_PLACEHOLDER_PREFIX + crypto.randomBytes(18).toString('hex')
}

function ensurePatchIdentity() {
  const state = loadPatchState()
  if (typeof state.placeholderApiKey !== 'string' || !state.placeholderApiKey) {
    state.placeholderApiKey = generatePlaceholderApiKey()
    state.createdAt = new Date().toISOString()
    savePatchState(state)
  }
  return state
}

function placeholderSummary(value) {
  if (!value) return '(missing)'
  return `${value.slice(0, 18)}...${value.slice(-6)}`
}

function patchSettings(gateway, registerGateway, identity) {
  backup(settingsPath)
  const settings = readJson(settingsPath)
  const oldApi = settings[CONFIG_KEY]
  const oldRegister = settings[REGISTER_CONFIG_KEY]
  const enabledAgents = typeof settings[ACP_CONFIG_KEY] === 'object' && settings[ACP_CONFIG_KEY] !== null
    ? settings[ACP_CONFIG_KEY]
    : {}
  settings[CONFIG_KEY] = gateway
  if (registerGateway) settings[REGISTER_CONFIG_KEY] = registerGateway
  settings[ACP_CONFIG_KEY] = { ...enabledAgents, 'devin-cloud': false }
  writeJson(settingsPath, settings)
  console.log(`settings ${CONFIG_KEY}: ${oldApi || DEFAULT_API_URL} -> ${gateway}`)
  if (registerGateway) console.log(`settings ${REGISTER_CONFIG_KEY}: ${oldRegister || DEFAULT_REGISTER_URL} -> ${registerGateway}`)
  console.log(`settings ${ACP_CONFIG_KEY}.devin-cloud -> false`)
  if (identity?.placeholderApiKey) {
    console.log(`gateway placeholder key: ${placeholderSummary(identity.placeholderApiKey)}`)
  }
}

function buildAuthSessionFallbackBlock(placeholderApiKey) {
  return `??{id:"windsurf-gateway",accessToken:"${placeholderApiKey}",account:{label:"Gateway",id:"windsurf-gateway"},scopes:[]}`
}

function patchExtension(gateway, registerGateway, identity) {
  if (!fs.existsSync(extensionPath)) {
    console.log(`extension not found: ${extensionPath}`)
    return false
  }
  backup(extensionPath)
  let content = fs.readFileSync(extensionPath, 'utf8')
  const before = content
  const placeholderApiKey = identity?.placeholderApiKey || LEGACY_GATEWAY_PLACEHOLDER_API_KEY
  content = content.replaceAll(`DEFAULT_API_SERVER_URL=\"${DEFAULT_API_URL}\"`, `DEFAULT_API_SERVER_URL=\"${gateway}\"`)
  content = content.replaceAll(`DEFAULT_API_SERVER_URL="${DEFAULT_API_URL}"`, `DEFAULT_API_SERVER_URL="${gateway}"`)
  if (registerGateway) {
    content = content.replaceAll(`DEFAULT_REGISTER_API_SERVER_URL=\"${DEFAULT_REGISTER_URL}\"`, `DEFAULT_REGISTER_API_SERVER_URL=\"${registerGateway}\"`)
    content = content.replaceAll(`DEFAULT_REGISTER_API_SERVER_URL="${DEFAULT_REGISTER_URL}"`, `DEFAULT_REGISTER_API_SERVER_URL="${registerGateway}"`)
  }
  if (AUTH_SESSION_FALLBACK_BLOCK_REGEX.test(content)) {
    content = content.replace(AUTH_SESSION_FALLBACK_BLOCK_REGEX, buildAuthSessionFallbackBlock(placeholderApiKey))
  } else {
    content = content.replace(AUTH_SESSION_FALLBACK_REGEX, (match) => {
      return `${match}${buildAuthSessionFallbackBlock(placeholderApiKey)}`
    })
  }
  if (!content.includes(USER_STATUS_FALLBACK_SENTINEL)) {
    content = content.replace(USER_STATUS_FALLBACK_REGEX, (_match, statusBarRef, authRef, _sessionRef, apiKeyRef) => {
      return `${statusBarRef}.StatusBar.getInstance().setAuthStatus(!0),${authRef}.windsurfAuth.setAuthStatus({apiKey:${apiKeyRef},allowedCommandModelConfigsProtoBinaryBase64:[],userStatusProtoBinaryBase64:""}),!0`
    })
  }
  if (content.includes(AUTH_SESSION_FALLBACK_MARKER) && !before.includes(AUTH_SESSION_FALLBACK_MARKER)) {
    console.log('extension auth bootstrap patched with gateway fallback session')
  }
  if (content.includes(USER_STATUS_FALLBACK_SENTINEL) && !before.includes(USER_STATUS_FALLBACK_SENTINEL)) {
    console.log('extension user-status fallback patched for gateway mode')
  }
  if (content === before) {
    console.log('extension already patched or patch markers not found')
    return false
  }
  fs.writeFileSync(extensionPath, content)
  console.log(`extension patched: ${extensionPath}`)
  console.log(`extension placeholder key: ${placeholderSummary(placeholderApiKey)}`)
  return true
}

function patchGlobalState(gateway, identity) {
  if (!fs.existsSync(globalStatePath)) {
    console.log(`globalState not found: ${globalStatePath}`)
    return false
  }
  let Database
  try {
    Database = require('better-sqlite3')
  } catch (e) {
    console.log('better-sqlite3 not installed; skip globalState')
    return false
  }
  backup(globalStatePath)
  const db = new Database(globalStatePath)
  
  // Patch API server URLs
  const rows = db.prepare('SELECT key,value FROM ItemTable WHERE key LIKE ? OR key LIKE ?').all('%apiServerUrl%', '%BASE_API_SERVER_URL%')
  for (const row of rows) {
    db.prepare('UPDATE ItemTable SET value = ? WHERE key = ?').run(JSON.stringify(gateway), row.key)
    console.log(`globalState ${row.key} -> ${gateway}`)
  }
  
  // Mock windsurfAuthStatus to skip signup
  const mockAuthStatus = {
    apiKey: identity?.placeholderApiKey || LEGACY_GATEWAY_PLACEHOLDER_API_KEY,
    allowedCommandModelConfigsProtoBinaryBase64: [],
    userStatusProtoBinaryBase64: ''
  }
  db.prepare(`
    INSERT INTO ItemTable (key, value) VALUES (?, ?)
    ON CONFLICT(key) DO UPDATE SET value = excluded.value
  `).run('windsurfAuthStatus', JSON.stringify(mockAuthStatus))
  console.log('globalState windsurfAuthStatus -> mocked')
  console.log(`globalState placeholder key: ${placeholderSummary(mockAuthStatus.apiKey)}`)
  
  // Mock onboarding completion
  db.prepare(`
    INSERT INTO ItemTable (key, value) VALUES (?, ?)
    ON CONFLICT(key) DO UPDATE SET value = excluded.value
  `).run('windsurfOnboarding', JSON.stringify(true))
  console.log('globalState windsurfOnboarding -> true')
  
  // Mock product education completed
  const mockProductEducation = {
    onboardingState: 2,
    onboardingItems: [
      {
        id: 'windsurf.prioritized.chat.open',
        title: 'Code with Cascade',
        completed: true,
        command: 'windsurf.prioritized.chat.open'
      }
    ]
  }
  db.prepare(`
    INSERT INTO ItemTable (key, value) VALUES (?, ?)
    ON CONFLICT(key) DO UPDATE SET value = excluded.value
  `).run('windsurfProductEducation', JSON.stringify(mockProductEducation))
  console.log('globalState windsurfProductEducation -> mocked')
  
  db.close()
  return rows.length > 0
}

function detect() {
  const state = loadPatchState()
  console.log(`configDir: ${configDir}`)
  console.log(`installDir: ${installDir}`)
  console.log(`settings: ${fs.existsSync(settingsPath) ? settingsPath : 'not found'}`)
  console.log(`globalState: ${fs.existsSync(globalStatePath) ? globalStatePath : 'not found'}`)
  console.log(`extension: ${fs.existsSync(extensionPath) ? extensionPath : 'not found'}`)
  console.log(`patchState: ${fs.existsSync(patchStateFile) ? patchStateFile : 'not found'}`)
  console.log(`placeholder key: ${placeholderSummary(state.placeholderApiKey)}`)
  console.log(`placeholder mode: ${state.placeholderApiKey === LEGACY_GATEWAY_PLACEHOLDER_API_KEY ? 'legacy-shared' : 'per-client'}`)
  if (fs.existsSync(settingsPath)) {
    const settings = readJson(settingsPath)
    console.log(`${CONFIG_KEY}: ${settings[CONFIG_KEY] || '(default)'}`)
    console.log(`${REGISTER_CONFIG_KEY}: ${settings[REGISTER_CONFIG_KEY] || '(default)'}`)
    console.log(`${ACP_CONFIG_KEY}.devin-cloud: ${settings[ACP_CONFIG_KEY]?.['devin-cloud']}`)
  }
  if (fs.existsSync(extensionPath)) {
    const content = fs.readFileSync(extensionPath, 'utf8')
    console.log(`contains ${DEFAULT_API_URL}: ${content.includes(DEFAULT_API_URL)}`)
    console.log(`contains ${DEFAULT_REGISTER_URL}: ${content.includes(DEFAULT_REGISTER_URL)}`)
    console.log(`contains auth-session fallback: ${AUTH_SESSION_FALLBACK_BLOCK_REGEX.test(content)}`)
    console.log(`contains user-status fallback: ${content.includes(USER_STATUS_FALLBACK_SENTINEL)}`)
  }
}

function restore() {
  if (!fs.existsSync(backupRoot)) {
    console.log('no backup root found')
    return
  }
  const dirs = fs.readdirSync(backupRoot).sort().reverse()
  if (dirs.length === 0) {
    console.log('no backups found')
    return
  }
  const dir = path.join(backupRoot, dirs[0])
  const files = fs.readdirSync(dir)
  for (const file of files) {
    const src = path.join(dir, file)
    if (file === 'settings.json') fs.copyFileSync(src, settingsPath)
    if (file === 'extension.js') fs.copyFileSync(src, extensionPath)
    if (file === 'state.vscdb') fs.copyFileSync(src, globalStatePath)
    console.log(`restored ${file}`)
  }
  console.log(`restored from ${dir}`)
  
  // Clean up mocked data from globalState
  let Database
  try {
    Database = require('better-sqlite3')
  } catch (e) {
    console.log('better-sqlite3 not installed; skip cleanup')
    return
  }
  
  if (fs.existsSync(globalStatePath)) {
    const db = new Database(globalStatePath)
    const mockKeys = ['windsurfAuthStatus', 'windsurfOnboarding', 'windsurfProductEducation']
    for (const key of mockKeys) {
      const row = db.prepare('SELECT key FROM ItemTable WHERE key = ?').get(key)
      if (row) {
        db.prepare('DELETE FROM ItemTable WHERE key = ?').run(key)
        console.log(`cleaned up ${key}`)
      }
    }
    db.close()
  }
}

function usage() {
  console.log('Usage:')
  console.log('  node patch.js detect')
  console.log('  node patch.js --gateway=https://gateway.example.com [--mode=config|extension|all]')
  console.log('  node patch.js --restore')
  console.log('Env:')
  console.log('  WINDSURF_GATEWAY_URL=https://gateway.example.com')
  console.log('  WINDSURF_CONFIG_DIR=/custom/config')
  console.log('  WINDSURF_INSTALL_DIR=/custom/resources/app')
}

function main() {
  banner()
  if (process.argv.includes('detect') || hasArg('--detect')) return detect()
  if (hasArg('--restore') || hasArg('-r')) return restore()
  const gateway = argValue('--gateway') || process.env.WINDSURF_GATEWAY_URL
  const registerGateway = argValue('--register-gateway') || process.env.WINDSURF_REGISTER_GATEWAY_URL
  const mode = argValue('--mode') || 'all'
  if (!gateway) return usage()
  if (!/^https?:\/\//.test(gateway)) throw new Error('gateway must start with http:// or https://')
  if (mode === 'config' || mode === 'all') {
    const identity = ensurePatchIdentity()
    patchSettings(gateway, registerGateway, identity)
    patchGlobalState(gateway, identity)
  }
  if (mode === 'extension' || mode === 'all') {
    const identity = ensurePatchIdentity()
    patchExtension(gateway, registerGateway, identity)
  }
  console.log('\npatch done. Restart Windsurf.')
}

main()

File: internal/patcher/patch.go
package patcher

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"windsurf-gateway/internal/gatewayuser"
)

var (
	authFallbackRegex          = regexp.MustCompile(`await i\.authentication\.getSession\(n\.WindsurfExtensionMetadata\.getInstance\(\)\.authProviderId,\[[\s\S]*?\],e\)`)
	authFallbackBlockRegex     = regexp.MustCompile(`\?\?\{id:"windsurf-gateway",accessToken:"([^"]+)",account:\{label:"Gateway",id:"windsurf-gateway"\},scopes:\[\]\}`)
	userStatusFallbackSentinel = `allowedCommandModelConfigsProtoBinaryBase64:[],userStatusProtoBinaryBase64:""}`
	userStatusFallbackRegex    = regexp.MustCompile(`([A-Za-z_$][\w$]*)\.StatusBar\.getInstance\(\)\.setAuthStatus\(!1\),([A-Za-z_$][\w$]*)\.windsurfAuth\.setAuthStatus\(null\),\(await\(0,([A-Za-z_$][\w$]*)\.getAuthSession\)\(\)\)\?\.accessToken===([A-Za-z_$][\w$]*)\|\|([A-Za-z_$][\w$]*)\.clearAuthentication\(\),!1`)
)

func Detect(configDir, installDir string) (DetectResult, error) {
	env := ResolveEnvironment(configDir, installDir)
	result := DetectResult{
		Environment: env,
		DetectedAt:  time.Now(),
	}

	state, err := loadPatchState(env.PatchState)
	if err == nil && strings.TrimSpace(state.PlaceholderAPIKey) != "" {
		result.PatchStateExists = true
		result.PatchStatePlaceholder = summarizeToken(state.PlaceholderAPIKey)
		result.PatchStateMode = classifyAuthMode(state.PlaceholderAPIKey)
	} else if err != nil && !errors.Is(err, os.ErrNotExist) {
		result.Messages = append(result.Messages, fmt.Sprintf("patch state unreadable: %v", err))
	}

	result.Settings = detectSettings(env.SettingsPath)
	result.GlobalState = detectGlobalState(env.GlobalState)
	result.Extension = detectExtension(env.ExtensionPath)
	return result, nil
}

func Apply(options ApplyOptions) (ApplyResult, error) {
	options.Mode = NormalizeMode(options.Mode)
	if strings.TrimSpace(options.GatewayURL) == "" {
		return ApplyResult{}, fmt.Errorf("gateway_url is required")
	}
	if !strings.HasPrefix(strings.ToLower(strings.TrimSpace(options.GatewayURL)), "http://") && !strings.HasPrefix(strings.ToLower(strings.TrimSpace(options.GatewayURL)), "https://") {
		return ApplyResult{}, fmt.Errorf("gateway_url must start with http:// or https://")
	}
	authToken := strings.TrimSpace(options.AuthToken)
	if !gatewayuser.IsToken(authToken) {
		return ApplyResult{}, fmt.Errorf("auth_token is required and must start with %s", gatewayuser.TokenPrefix)
	}

	env := ResolveEnvironment(options.ConfigDir, options.InstallDir)
	backup := newBackupSession(env.BackupRoot)
	_ = backup.Backup(env.PatchState)
	effectiveToken := authToken
	effectiveMode := "gateway-user"

	messages := make([]string, 0, 8)
	if options.Mode == ModeConfig || options.Mode == ModeAll {
		if err := backup.Backup(env.SettingsPath); err != nil {
			return ApplyResult{}, err
		}
		if err := patchSettings(env.SettingsPath, options.GatewayURL, options.RegisterGatewayURL); err != nil {
			return ApplyResult{}, err
		}
		messages = append(messages, "settings.json updated")

		if err := backup.Backup(env.GlobalState); err != nil {
			return ApplyResult{}, err
		}
		patched, err := patchGlobalState(env.GlobalState, options.GatewayURL, effectiveToken)
		if err != nil {
			return ApplyResult{}, err
		}
		if patched {
			messages = append(messages, "state.vscdb updated")
		} else {
			messages = append(messages, "state.vscdb not found; skipped")
		}
	}

	if options.Mode == ModeExtension || options.Mode == ModeAll {
		if err := backup.Backup(env.ExtensionPath); err != nil {
			return ApplyResult{}, err
		}
		patched, err := patchExtension(env.ExtensionPath, options.GatewayURL, options.RegisterGatewayURL, effectiveToken)
		if err != nil {
			return ApplyResult{}, err
		}
		if patched {
			messages = append(messages, "extension.js updated")
		} else {
			messages = append(messages, "extension.js not found or already patched")
		}
	}

	detect, err := Detect(options.ConfigDir, options.InstallDir)
	if err != nil {
		return ApplyResult{}, err
	}
	result := ApplyResult{
		BackupDir:             backup.dir,
		EffectiveTokenSummary: summarizeToken(effectiveToken),
		EffectiveAuthMode:     effectiveMode,
		Detect:                detect,
		Messages:              messages,
	}
	if len(backup.copied) == 0 {
		result.BackupDir = ""
	}
	return result, nil
}

func Restore(configDir, installDir string) (RestoreResult, error) {
	env := ResolveEnvironment(configDir, installDir)
	latest, err := latestBackupDir(env.BackupRoot)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return RestoreResult{}, fmt.Errorf("no backup found")
		}
		return RestoreResult{}, err
	}

	messages := make([]string, 0, 4)
	for _, item := range []struct {
		name   string
		target string
	}{
		{name: filepath.Base(env.SettingsPath), target: env.SettingsPath},
		{name: filepath.Base(env.ExtensionPath), target: env.ExtensionPath},
		{name: filepath.Base(env.GlobalState), target: env.GlobalState},
		{name: filepath.Base(env.PatchState), target: env.PatchState},
	} {
		source := filepath.Join(latest, item.name)
		data, readErr := os.ReadFile(source)
		if readErr != nil {
			continue
		}
		if err := os.MkdirAll(filepath.Dir(item.target), 0o755); err != nil {
			return RestoreResult{}, err
		}
		if err := os.WriteFile(item.target, data, 0o644); err != nil {
			return RestoreResult{}, err
		}
		messages = append(messages, fmt.Sprintf("restored %s", item.name))
	}

	if cleaned, err := cleanupMockedGlobalState(env.GlobalState); err != nil {
		messages = append(messages, fmt.Sprintf("global state cleanup failed: %v", err))
	} else if cleaned > 0 {
		messages = append(messages, fmt.Sprintf("cleaned %d mocked global-state keys", cleaned))
	}

	detect, err := Detect(configDir, installDir)
	if err != nil {
		return RestoreResult{}, err
	}
	return RestoreResult{RestoredFrom: latest, Detect: detect, Messages: messages}, nil
}

func patchSettings(settingsPath, gatewayURL, registerGatewayURL string) error {
	settings := map[string]any{}
	if data, err := os.ReadFile(settingsPath); err == nil && len(data) > 0 {
		if err := json.Unmarshal(data, &settings); err != nil {
			return fmt.Errorf("parse settings.json: %w", err)
		}
	} else if err != nil && !os.IsNotExist(err) {
		return err
	}

	settings[ConfigKey] = gatewayURL
	if strings.TrimSpace(registerGatewayURL) != "" {
		settings[RegisterConfigKey] = registerGatewayURL
	}
	enabledAgents := map[string]any{}
	if existing, ok := settings[ACPConfigKey].(map[string]any); ok {
		for key, value := range existing {
			enabledAgents[key] = value
		}
	}
	enabledAgents["devin-cloud"] = false
	settings[ACPConfigKey] = enabledAgents
	return writeJSON(settingsPath, settings)
}

func patchExtension(extensionPath, gatewayURL, registerGatewayURL, effectiveToken string) (bool, error) {
	data, err := os.ReadFile(extensionPath)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}

	content := string(data)
	before := content
	content = replaceExtensionConstant(content, "DEFAULT_API_SERVER_URL", DefaultAPIURL, gatewayURL)
	if strings.TrimSpace(registerGatewayURL) != "" {
		content = replaceExtensionConstant(content, "DEFAULT_REGISTER_API_SERVER_URL", DefaultRegisterURL, registerGatewayURL)
	}

	fallbackBlock := buildAuthFallbackBlock(effectiveToken)
	if authFallbackBlockRegex.MatchString(content) {
		content = authFallbackBlockRegex.ReplaceAllStringFunc(content, func(string) string {
			return fallbackBlock
		})
	} else {
		content = authFallbackRegex.ReplaceAllStringFunc(content, func(match string) string {
			return match + fallbackBlock
		})
	}
	if !strings.Contains(content, userStatusFallbackSentinel) {
		content = userStatusFallbackRegex.ReplaceAllString(content, `${1}.StatusBar.getInstance().setAuthStatus(!0),${2}.windsurfAuth.setAuthStatus({apiKey:${4},allowedCommandModelConfigsProtoBinaryBase64:[],userStatusProtoBinaryBase64:""}),!0`)
	}

	if content == before {
		return false, nil
	}
	return true, os.WriteFile(extensionPath, []byte(content), 0o644)
}

func buildAuthFallbackBlock(token string) string {
	return fmt.Sprintf(`??{id:"windsurf-gateway",accessToken:"%s",account:{label:"Gateway",id:"windsurf-gateway"},scopes:[]}`, token)
}

func replaceExtensionConstant(content, key, oldValue, newValue string) string {
	quoted := key + `="` + oldValue + `"`
	replaced := key + `="` + newValue + `"`
	escapedQuoted := key + `=\"` + oldValue + `\"`
	escapedReplaced := key + `=\"` + newValue + `\"`
	content = strings.ReplaceAll(content, quoted, replaced)
	content = strings.ReplaceAll(content, escapedQuoted, escapedReplaced)
	return content
}

func detectSettings(settingsPath string) SettingsInfo {
	info := SettingsInfo{}
	data, err := os.ReadFile(settingsPath)
	if err != nil {
		return info
	}
	info.Exists = true
	var settings map[string]any
	if jsonErr := json.Unmarshal(data, &settings); jsonErr != nil {
		return info
	}
	if gateway, ok := settings[ConfigKey].(string); ok {
		info.GatewayURL = gateway
	}
	if register, ok := settings[RegisterConfigKey].(string); ok {
		info.RegisterGatewayURL = register
	}
	if enabledAgents, ok := settings[ACPConfigKey].(map[string]any); ok {
		if value, ok := enabledAgents["devin-cloud"].(bool); ok {
			info.DevinCloudDisabled = !value
		}
	}
	return info
}

func detectExtension(extensionPath string) ExtensionInfo {
	info := ExtensionInfo{}
	data, err := os.ReadFile(extensionPath)
	if err != nil {
		if !os.IsNotExist(err) {
			info.ReadError = err.Error()
		}
		return info
	}
	info.Exists = true
	content := string(data)
	info.ContainsDefaultAPI = strings.Contains(content, DefaultAPIURL)
	info.ContainsDefaultRegister = strings.Contains(content, DefaultRegisterURL)
	info.HasAuthFallback = authFallbackBlockRegex.MatchString(content)
	info.HasUserStatusFallback = strings.Contains(content, userStatusFallbackSentinel)
	matches := authFallbackBlockRegex.FindStringSubmatch(content)
	if len(matches) > 1 {
		info.AuthTokenSummary = summarizeToken(matches[1])
		info.AuthMode = classifyAuthMode(matches[1])
	}
	return info
}
